### PR TITLE
send Varnish BAN requests with smaller headers

### PR DESCRIPTION
**Pull request by jocel1**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| License       | MIT

The idea of this PR is to split the iris in several smaller chunks to avoid the "400 Request Header Or Cookie Too Large" on Varnish when the iris contains a large amount of elements.
CHUNK_SIZE is currently hardcoded, but if this could be controlled as a YAML config parameter, it would be even better.

### CODE REVIEW

**Comment by soyuka** [comment]
Is there a limit on varnish? How did you come up with the value `100`? 

Nice improvement!

**Comment by bastnic** [comment]
FYI, there is a chunk made in FosHTTPCache https://github.com/FriendsOfSymfony/FOSHttpCache/blob/master/src/ProxyClient/Varnish.php#L84 and there is méthod dedicated to calculate how much tags fits in the http header : https://github.com/FriendsOfSymfony/FOSHttpCache/blob/master/src/ProxyClient/HttpProxyClient.php#L137

**Comment by jocel1** [comment]
@soyuka it's a really conservative value, but default values in Varnish are not that big (you have by default http_req_hdr_len=8k and http_req_size=32k in Varnish, but even with http_req_size=64k & http_req_hdr_len=4096k I was still experiencing 400 Bad Request issues).

**Comment by jocel1** [comment]
@bastnic thanks for the tips, this looks great!

**Comment by dunglas** [comment]
👍 on my side! Copying (and crediting) the method from FOSHttpCache looks like a good idea. Could you also fix the PHPStan issue please?

**Comment by jocel1** [comment]
@dunglas changes done + adding a yaml config variable to set the max_header_length supported by the server

**Comment by jocel1** [comment]
Hi @dunglas , any plan to merge this one ? :)

File: src/HttpCache/VarnishPurger.php [hunk 1]
```diff
@@ -24,14 +24,43 @@
...
         $this->clients = $clients;
+        $this->maxHeaderLength = $maxHeaderLength;
+    }
+
+    /**
+     * Calculate how many tags fit into the header.
+     *
+     * This assumes that the tags are separated by one character.
+     *
+     * From https://github.com/FriendsOfSymfony/FOSHttpCache/blob/master/src/ProxyClient/HttpProxyClient.php#L137
```
Comment (teohhanhui) [inline src/HttpCache/VarnishPurger.php]: Please reference a specific commit id, otherwise this comment will be outdated real soon.

Comment (jocel1) [reply src/HttpCache/VarnishPurger.php]: right, done!

**Comment by Cocray** [comment]
> Hi @dunglas , any plan to merge this one ? :)

Hello there! Should we have a release date for this feature? It's something we're waiting with excitement :)

File: src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php [hunk 2]
```diff
@@ -375,6 +375,10 @@ private function addHttpCacheSection(ArrayNodeDefinition $rootNode): void
                                     ->prototype('scalar')->end()
                                     ->info('URLs of the Varnish servers to purge using cache tags when a resource is updated.')
                                 ->end()
+                                ->integerNode('max_header_length')
+                                    ->defaultValue(7500)
```
Comment (teohhanhui) [inline src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php]: How is this default derived?

File: src/HttpCache/VarnishPurger.php [hunk 3]
```diff
@@ -24,14 +24,43 @@
...
 {
+    private $maxHeaderLength;
     private $clients;
 
     /**
      * @param ClientInterface[] $clients
+     * @param int               $maxHeaderLength
      */
-    public function __construct(array $clients)
+    public function __construct(array $clients, $maxHeaderLength = 7500)
```
Comment (teohhanhui) [inline src/HttpCache/VarnishPurger.php]:
```suggestion
    public function __construct(array $clients, int $maxHeaderLength = 7500)
```

File: src/HttpCache/VarnishPurger.php [hunk 4]
```diff
@@ -24,14 +24,43 @@
  */
 final class VarnishPurger implements PurgerInterface
 {
+    private $maxHeaderLength;
     private $clients;
 
     /**
      * @param ClientInterface[] $clients
+     * @param int               $maxHeaderLength
```
Comment (teohhanhui) [inline src/HttpCache/VarnishPurger.php]: Remove useless phpdoc (use typehint instead).

File: src/HttpCache/VarnishPurger.php [hunk 5]
```diff
@@ -24,14 +24,43 @@
  */
 final class VarnishPurger implements PurgerInterface
 {
+    private $maxHeaderLength;
```
Comment (teohhanhui) [inline src/HttpCache/VarnishPurger.php]: This should be moved after `$clients`.

File: src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php (hunk 2 above)
Comment (jocel1) [reply src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php]: Default Apache max headers size is 8K so the idea is to let some room for other headers, hence the 7500 value

File: src/HttpCache/VarnishPurger.php (hunk 4 above)
Comment (jocel1) [reply src/HttpCache/VarnishPurger.php]: ok for the typehint, but keeping the phpdoc avoid warning from IDE like phpstorm

Comment (jocel1) [reply src/HttpCache/VarnishPurger.php]: phpdoc removed anyway :)

**Comment by jocel1** [comment]
Hi there! Any update on this one?

File: src/HttpCache/VarnishPurger.php [hunk 6]
```diff
@@ -25,13 +25,41 @@
...
+     * This assumes that the tags are separated by one character.
+     *
+     * From https://github.com/FriendsOfSymfony/FOSHttpCache/blob/2.8.0/src/ProxyClient/HttpProxyClient.php#L137
+     *
+     * @param string[] $escapedTags
+     * @param string   $glue        The concatenation string to use
+     *
+     * @return int Number of tags per tag invalidation request
+     */
+    private function determineTagsPerHeader($escapedTags, $glue)
```
Comment (teohhanhui) [inline src/HttpCache/VarnishPurger.php]:
```suggestion
    private function determineTagsPerHeader(array $escapedTags, string $glue): int
```

**Comment by teohhanhui** [comment]
phpstan failure not related.

**Comment by teohhanhui** [comment]
Thanks @jocel1! :tada: :rocket:

**Pull request merged**